From 5d9f0d8333c4b507c4431f343e218a166c16f983 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Veronika=20Hanul=C3=ADkov=C3=A1?= <vhanulik@redhat.com>
Date: Tue, 2 Jul 2024 10:50:34 +0200
Subject: [PATCH] Allow OpenSSL engine header only for <3.0

---
 src/pkcs11/openssl.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/pkcs11/openssl.c b/src/pkcs11/openssl.c
index 9a36044ff..c9c7c7150 100644
--- a/src/pkcs11/openssl.c
+++ b/src/pkcs11/openssl.c
@@ -33,6 +33,7 @@
 #include <openssl/conf.h>
 #include <openssl/opensslconf.h> /* for OPENSSL_NO_* */
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
+#define OPENSSL_NO_ENGINE
 #include <openssl/core_names.h>
 #include <openssl/param_build.h>
 #endif
@@ -258,7 +259,6 @@ sc_pkcs11_register_openssl_mechanisms(struct sc_pkcs11_card *p11card)
  * Engine support is being deprecated in 3.0. OpenSC loads GOST as engine.
  * When GOST developers convert to provider, we can load the provider
  */
-#if OPENSSL_VERSION_NUMBER < 0x30000000L
 #if !defined(OPENSSL_NO_ENGINE)
 	ENGINE *e;
 /* crypto locking removed in 1.1 */
@@ -304,7 +304,6 @@ sc_pkcs11_register_openssl_mechanisms(struct sc_pkcs11_card *p11card)
 		CRYPTO_set_locking_callback(locking_cb);
 #endif
 #endif /* !defined(OPENSSL_NO_ENGINE) */
-#endif /* OPENSSL_VERSION_NUMBER < 0x30000000L */
 
 	openssl_sha1_mech.mech_data = sc_evp_md(context, "sha1");
 	openssl_sha1_mech.free_mech_data = ossl_md_free;
-- 
2.45.2

