diff -up opensc-0.11.6/src/libopensc/libopensc.pc.in.config opensc-0.11.6/src/libopensc/libopensc.pc.in
--- opensc-0.11.6/src/libopensc/libopensc.pc.in.config	2008-07-31 14:20:07.000000000 +0200
+++ opensc-0.11.6/src/libopensc/libopensc.pc.in	2008-09-02 14:49:04.000000000 +0200
@@ -6,6 +6,6 @@ includedir=@includedir@
 Name: libopensc
 Description: libopensc
 Version: @VERSION@
-Libs: -L${libdir} -lopensc -lscconf
+Libs: -L${libdir} -lopensc
 Cflags: -I${includedir}
 
diff -up opensc-0.11.6/src/libopensc/opensc-config.in.config opensc-0.11.6/src/libopensc/opensc-config.in
--- opensc-0.11.6/src/libopensc/opensc-config.in.config	2008-07-31 14:20:07.000000000 +0200
+++ opensc-0.11.6/src/libopensc/opensc-config.in	2008-09-02 14:52:12.000000000 +0200
@@ -15,10 +15,10 @@ prefix="@prefix@"
 exec_prefix="@exec_prefix@"
 exec_prefix_set=no
 
-libdir="@libdir@"
-includedir="@includedir@"
+libdir=""
+includedir=""
 opensc_cflags=""
-opensc_libs="-L${libdir} -lopensc"
+opensc_libs="-lopensc"
 
 usage()
 {
