diff -ur opensc-0.13.0.orig/src/pkcs15init/myeid.profile opensc-0.13.0/src/pkcs15init/myeid.profile
--- opensc-0.13.0.orig/src/pkcs15init/myeid.profile	2012-12-04 15:43:40.000000000 +0100
+++ opensc-0.13.0/src/pkcs15init/myeid.profile	2014-01-06 14:04:22.795401836 +0100
@@ -201,7 +201,7 @@
                     acl       = READ=$PIN, UPDATE=$PIN, DELETE=$PIN;
                 }
                 EF data {
-                    file-id   = 4501;
+                    file-id   = 4601;
                     structure = transparent;
                     acl       = READ=NONE, UPDATE=$PIN, DELETE=$PIN;
                 }
