Index: src/pam/pam_support.c
===================================================================
--- src/pam/pam_support.c	(revision 2139)
+++ src/pam/pam_support.c	(revision 2140)
@@ -205,7 +205,8 @@
 
 static void _cleanup(pam_handle_t * pamh, void *x, int error_status)
 {
-	_pam_delete((char *) x);
+	char *y = (char *) x;
+	_pam_delete(y);
 }
 
 /* ************************************************************** *
